sap-vcap-services/src/main/java/com/sap/cloud/environment/servicebinding/SapVcapServicesServiceBindingAccessor.java

```diff
@@ -101,45 +105,72 @@ public List<ServiceBinding> getServiceBindings()
-        }
-        catch( final JSONException e ) {
-            logger.debug("Skipping '{}': Unexpected format.", VCAP_SERVICES);
+        final Object serviceBindings = vcapServices.get(serviceName);
+        if( !(serviceBindings instanceof JSONArray) ) {
+            logger.debug("Skipping '{}': Unexpected format.", serviceName);
             return Collections.emptyList();
         }
 
-        final List<ServiceBinding> serviceBindings = new ArrayList<>(jsonServiceBindings.length());
-        for( int i = 0; i < jsonServiceBindings.length(); ++i ) {
-            final JSONObject jsonServiceBinding;
-            try {
-                jsonServiceBinding = jsonServiceBindings.getJSONObject(i);
-            }
-            catch( final JSONException e ) {
-                continue;
-            }
-
-            serviceBindings.add(toServiceBinding(jsonServiceBinding, serviceName));
+        final List<ServiceBinding> result =
+            StreamSupport
+                .stream(((Iterable<?>) serviceBindings).spliterator(), false)
+                .filter(JSONObject.class::isInstance)
+                .flatMap(binding -> toServiceBindings(((JSONObject) binding).toMap(), serviceName).stream())
+                .collect(Collectors.toList());
+
+        logger.debug("Successfully read {} service binding(s) from '{}'.", result.size(), serviceName);
+        return result;
+    }
+
+    @Nonnull
+    protected
+        Collection<ServiceBinding>
+        toServiceBindings( @Nonnull final Map<String, Object> props, @Nonnull final String serviceName )
+    {
+        if( isUserProvided(serviceName) ) {
+            final List<String> tags = getTagsFromProperties(props);
+            return tags.stream().map(tag -> createServiceBinding(props, serviceName, tag)).collect(Collectors.toList());
+        } else {
+            return Collections.singleton(createServiceBinding(props, serviceName, serviceName));
         }
+    }
 
-        logger.debug("Successfully read {} service binding(s) from '{}'.", serviceBindings.size(), VCAP_SERVICES);
-        return serviceBindings;
+    protected boolean isUserProvided( @Nonnull final String serviceName )
+    {
+        return "user-provided".equalsIgnoreCase(serviceName);
     }
 
+    @SuppressWarnings( "unchecked" )
     @Nonnull
-    private
-        ServiceBinding
-        toServiceBinding( @Nonnull final JSONObject jsonServiceBinding, @Nonnull final String serviceName )
+    protected List<String> getTagsFromProperties( @Nonnull final Map<String, Object> properties )
+    {
+        final Object tags = properties.get("tags");
+        if( !(tags instanceof List) ) {
+            logger.debug("No \"tags\" found in service binding {}.", tags);
+            return Collections.emptyList();
+        }
+        if( ((List<?>) tags).isEmpty() || !(((List<?>) tags).get(0) instanceof String) ) {
+            logger.debug("Empty or unexpected format for \"tags\" in service binding {}.", tags);
```

Comment:
This is slightly misleading as I would have expected the service binding to be logged.

```suggestion
            logger.debug("Empty or unexpected format for \"tags\" in service binding: {}.", tags);
```